app/assets/javascripts/discourse/controllers/preferences/second-factor.js.es6

```diff
@@ -44,55 +30,61 @@ export default Ember.Controller.extend({
     return user && user.get("enforcedSecondFactor");
   },
 
-  toggleSecondFactor(enable) {
-    if (!this.secondFactorToken) return;
+  handleError(error) {
+    if (error.jqXHR) {
+      error = error.jqXHR;
+    }
+    let parsedJSON = error.responseJSON;
+    if (parsedJSON.error_type === "invalid_access") {
+      const usernameLower = this.get("model").username.toLowerCase();
+      DiscourseURL.redirectTo(
+        userPath(`${usernameLower}/preferences/second-factor`)
+      );
+    } else {
+      popupAjaxError(error);
+    }
+  },
+
+  loadSecondFactors() {
+    if (this.get("dirty") === false) {
+      return;
+    }
     this.set("loading", true);
 
-    this.model
-      .toggleSecondFactor(
-        this.secondFactorToken,
-        this.secondFactorMethod,
-        SECOND_FACTOR_METHODS.TOTP,
-        enable
-      )
+    this.get("model")
+      .loadSecondFactorCodes(this.get("password"))
       .then(response => {
         if (response.error) {
           this.set("errorMessage", response.error);
           return;
         }
 
-        this.set("errorMessage", null);
-        DiscourseURL.redirectTo(
-          userPath(`${this.model.username.toLowerCase()}/preferences`)
+        this.setProperties({
+          errorMessage: null,
+          loaded: true,
+          totps: response.totps,
+          password: null,
+          dirty: false
+        });
+        this.get("model").set(
```

Comment:
this.set("model. second_factor_enabled", response.totps && response.totps.length > 0) would work?